rocAL/source/decoders/image/rocjpeg_decoder.cpp

```diff
@@ -0,0 +1,386 @@
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+
+#include "pipeline/commons.h"
+#include <stdio.h>
+#include <string.h>
+#include "decoders/image/rocjpeg_decoder.h"
+
+#if ENABLE_ROCJPEG
+
+#include "hip/hip_runtime_api.h"
+#include "hip/hip_runtime.h"
+#include "rocal_hip_kernels.h"
+
+#define CHECK_HIP(call) {                                             \
+    hipError_t hip_status = (call);                                   \
+    if (hip_status != hipSuccess) {                                   \
+        std::cerr << "HIP failure: 'status: " << hipGetErrorName(hip_status) << "' at " << __FILE__ << ":" << __LINE__ << std::endl;\
+        exit(1);                                                      \
+    }                                                                 \
+}
+
+#define CHECK_ROCJPEG(call) {                                             \
+    RocJpegStatus rocjpeg_status = (call);                                \
+    if (rocjpeg_status != ROCJPEG_STATUS_SUCCESS) {                       \
+        std::cerr << #call << " returned " << rocJpegGetErrorName(rocjpeg_status) << " at " <<  __FILE__ << ":" << __LINE__ << std::endl;\
+        exit(1);                                                        \
+    }                                                                     \
+}
+
+HWRocJpegDecoder::HWRocJpegDecoder() {
+};
+
+/**
+ * @brief Gets the channel pitch and sizes.
+ *
+ * This function gets the channel pitch and sizes based on the specified output format, chroma subsampling,
+ * output image, and channel sizes.
+ *
+ * @param output_format The output format.
+ * @param subsampling The chroma subsampling.
+ * @param widths The array to store the channel widths.
+ * @param heights The array to store the channel heights.
+ * @param num_channels The number of channels.
+ * @param output_image The output image.
+ * @param channel_sizes The array to store the channel sizes.
+ * @return The channel pitch.
+ */
+int GetChannelPitchAndSizes(RocJpegOutputFormat output_format, RocJpegChromaSubsampling subsampling, uint32_t *widths, uint32_t *heights,
+                            uint32_t &num_channels, RocJpegImage &output_image, uint32_t *channel_sizes) {
+    switch (output_format) {
+        case ROCJPEG_OUTPUT_NATIVE:
+            switch (subsampling) {
+                case ROCJPEG_CSS_444:
+                    num_channels = 3;
+                    output_image.pitch[2] = output_image.pitch[1] = output_image.pitch[0] = widths[0];
+                    channel_sizes[2] = channel_sizes[1] = channel_sizes[0] = output_image.pitch[0] * heights[0];
```

Comment:
As I mentioned earlier, please refer to the [GetChannelPitchAndSizes](https://github.com/ROCm/rocJPEG/blob/develop/samples/rocjpeg_samples_utils.h#L318) function in rocJPEG for guidance. You will need to calculate the pitch and channel sizes depending on whether a Region of Interest (ROI) is present. Therefore, the two lines above should be modified accordingly.

```
output_image.pitch[2] = output_image.pitch[1] = output_image.pitch[0] = is_roi_valid ? roi_width : widths[0];
channel_sizes[2] = channel_sizes[1] = channel_sizes[0] = align(output_image.pitch[0] * (is_roi_valid ? roi_height : heights[0]), mem_alignment);

```

Reply:
Please see my comment above.

Reply:
Done